src/module_4/push_exploration_non_linear.md

```diff
@@ -0,0 +1,2005 @@
+    """
+    Return a df with X set (features) and a series y set (outcome)
+
+    Parameters:
+    - df: pd.DataFrame
+
+    Returns:
+    - X: pd.DataFrame (features only)
+    - y: pd.Series (label to predict)
+    """
+    X = df.drop(columns=label)
+    y = df[label]
+
+    return X, y
+
+
+def temporal_data_split(
+    data: pd.DataFrame,
+    validation_days: int = 10,
+    test_days: int = 10,
+    label: str = "outcome",
+    cols: List[str] = None,
+):
+    """
+    Perform the temporal data splitting into train, validation, and test set.
+
+    Parameters:
+    - data : pd.DataFrame
+    - validation_days (int): Number of days for the validation set (default: 10).
+    - test_days (int): Number of days for the test set (default: 10).
+    - label (str): Name of the outcome variable (default: 'outcome').
+
+    Returns:
+    - pd.DataFrame: Training set features.
+    - pd.Series: Training set outcome.
+    - pd.DataFrame: Validation set features.
+    - pd.Series: Validation set outcome.
+    - pd.DataFrame: Test set features.
+    - pd.Series: Test set outcome.
+    """
+    try:
+        # Check order_date is a datetime and then we group the orders
+        try:
+            data["order_date"] = pd.to_datetime(data["order_date"]).dt.date
+        except KeyError as ke:
+            logging.error(f"Key Error: {str(ke)}")
+            return
+        daily_orders = data.groupby("order_date").order_id.nunique()
+
+        start_date = daily_orders.index.min()
+        end_date = daily_orders.index.max()
+        logging.info(f"Date from: {start_date}")
+        logging.info(f"Date to: {end_date}")
+
+        # Based on the number of days, we get the train days
+        num_days = len(daily_orders)
+        train_days = num_days - test_days - validation_days
+
+        # Train
+        train_start = daily_orders.index.min()
+        train_end = train_start + timedelta(days=train_days)
+
+        # Validation (no need to define test)
+        validation_end = train_end + timedelta(days=validation_days)
+
+        if cols is None:
+            # Use all columns if cols is not specified
+            cols = data.columns.tolist()
+
+        train = data[data.order_date <= train_end][cols]
+        val = data[(data.order_date > train_end) & (data.order_date <= validation_end)][
+            cols
+        ]
+        test = data[data.order_date > validation_end][cols]
+
+        logging.info(f"Train set ratio: {len(train) / len(data):.2%}")
+        logging.info(f"Validation set ratio: {len(val) / len(data):.2%}")
+        logging.info(f"Test set ratio: {len(test) / len(data):.2%}")
+
+        X_train, y_train = split_sets(train, label)
+        X_val, y_val = split_sets(val, label)
+        X_test, y_test = split_sets(test, label)
+
+        return X_train, y_train, X_val, y_val, X_test, y_test
+
+    # not the best practice to catch all exceptio
+    except KeyError as ke:
+        logging.error(f"Key Error: {str(ke)}")
+        return
```

Comment:
I thought you would do a moving train / val moving window in order to evaluate the model over different periods but I don't see you do this